File: tests/python/test_operator_pools.py
import cudaq_algorithms as algorithms


def _make_uccsd_operator_pool(num_qubits, num_electrons, spin=0):
    return algorithms.stateprep.make_uccsd_operator_pool(
        num_qubits, num_electrons, spin)


def _assert_max_qubit_count(operators, max_qubits):
    for op in operators:
        assert op.qubit_count <= max_qubits


def test_generate_with_default_config():
    operators = _make_uccsd_operator_pool(num_qubits=4, num_electrons=2)
    assert operators
    assert len(operators) == 2 + 1
    _assert_max_qubit_count(operators, 4)


def test_generate_with_custom_coefficients():
    operators = _make_uccsd_operator_pool(num_qubits=4, num_electrons=2)

    assert operators
    assert len(operators) == 2 + 1

    for op in operators:
        assert op.qubit_count <= 4
        expected_coefficients = [0.5, 0.125]
        for term in op:
            assert abs(
                term.evaluate_coefficient().real) in expected_coefficients


def test_generate_with_odd_electrons():
    operators = _make_uccsd_operator_pool(num_qubits=6,
                                          num_electrons=3,
                                          spin=1)

    assert operators
    assert len(operators) == 2 * 2 + 4
    _assert_max_qubit_count(operators, 6)


def test_generate_with_large_system():
    operators = _make_uccsd_operator_pool(num_qubits=20, num_electrons=10)

    assert operators
    assert len(operators) == 875
    _assert_max_qubit_count(operators, 20)


def test_uccgsd_operator_pool_counts():
    assert len(algorithms.stateprep.make_uccgsd_operator_pool(4)) == 9
    assert len(algorithms.stateprep.make_uccgsd_operator_pool(4, True,
                                                              False)) == 6
    assert len(algorithms.stateprep.make_uccgsd_operator_pool(4, False,
                                                              True)) == 3

    operators = algorithms.stateprep.make_uccgsd_operator_pool(8)
    assert len(operators) == 238
    _assert_max_qubit_count(operators, 8)


def test_upccgsd_operator_pool_counts():
    operators = algorithms.stateprep.make_upccgsd_operator_pool(20)
    doubles = algorithms.stateprep.make_upccgsd_operator_pool(20, True)

    assert len(operators) == 135
    assert len(doubles) == 45
    _assert_max_qubit_count(operators, 20)
    _assert_max_qubit_count(doubles, 20)


def test_ceo_operator_pool_counts():
    two_orbital_operators = algorithms.stateprep.make_ceo_operator_pool(2)
    four_orbital_operators = algorithms.stateprep.make_ceo_operator_pool(4)

    assert len(two_orbital_operators) == 4
    assert len(four_orbital_operators) == 96
    _assert_max_qubit_count(two_orbital_operators, 4)
    _assert_max_qubit_count(four_orbital_operators, 8)


def test_uccsd_operator_pool_correctness():
    pool = _make_uccsd_operator_pool(num_qubits=4, num_electrons=2)

    generated = [[(term.get_pauli_word(4), term.evaluate_coefficient())
                  for term in op] for op in pool]

    expected_operators = [["XZYI", "YZXI"], ["IXZY", "IYZX"],
                          [
                              "YYYX", "YXXX", "XXYX", "YYXY", "XYYY", "XXXY",
                              "YXYY", "XYXX"
                          ]]
    expected_coefficients = [[complex(-0.5, 0),
                              complex(0.5, 0)],
                             [complex(-0.5, 0),
                              complex(0.5, 0)],
                             [
                                 complex(-0.125, 0),
                                 complex(-0.125, 0),
                                 complex(0.125, 0),
                                 complex(-0.125, 0),
                                 complex(0.125, 0),
                                 complex(0.125, 0),
                                 complex(0.125, 0),
                                 complex(-0.125, 0)
                             ]]

    assert len(generated) == len(expected_operators)

    valid_chars = set("IXYZ")
    for i, operator_terms in enumerate(generated):
        for pauli_word, coefficient in operator_terms:
            assert len(pauli_word) <= 4
            assert set(pauli_word).issubset(valid_chars)

            expected_index = expected_operators[i].index(pauli_word)
            assert coefficient == expected_coefficients[i][expected_index]

File: lib/unittests/test_operator_pools.cpp
/*******************************************************************************
 * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "cudaq/algorithms/stateprep/excitations.h"

#include <cmath>
#include <complex>
#include <gtest/gtest.h>
#include <set>
#include <vector>

namespace stateprep = cudaq::algorithms::stateprep;

namespace {

void expect_max_degree_less_than(const std::vector<cudaq::spin_op> &operators,
                                 std::size_t max_degree) {
  for (const auto &op : operators)
    EXPECT_LT(op.max_degree(), max_degree);
}

bool is_expected_uccsd_coefficient(std::complex<double> coefficient) {
  const auto real = coefficient.real();
  return std::abs(coefficient.imag()) < 1e-12 &&
         (std::abs(std::abs(real) - 0.5) < 1e-12 ||
          std::abs(std::abs(real) - 0.125) < 1e-12);
}

} // namespace

TEST(UCCSDOperatorPool, GenerateWithDefaultConfig) {
  auto operators = stateprep::make_uccsd_operator_pool(4, 2);

  ASSERT_FALSE(operators.empty());
  EXPECT_EQ(operators.size(), 3);
  expect_max_degree_less_than(operators, 4);
}

TEST(UCCSDOperatorPool, GenerateWithCustomCoefficients) {
  auto operators = stateprep::make_uccsd_operator_pool(4, 2);

  ASSERT_FALSE(operators.empty());
  EXPECT_EQ(operators.size(), 3);

  for (const auto &op : operators) {
    EXPECT_LT(op.max_degree(), 4);
    for (const auto &term : op)
      EXPECT_TRUE(is_expected_uccsd_coefficient(term.evaluate_coefficient()));
  }
}

TEST(UCCSDOperatorPool, GenerateWithOddElectrons) {
  auto operators = stateprep::make_uccsd_operator_pool(6, 3, 1);

  ASSERT_FALSE(operators.empty());
  EXPECT_EQ(operators.size(), 8);
  expect_max_degree_less_than(operators, 6);
}

TEST(UCCSDOperatorPool, GenerateWithLargeSystem) {
  auto operators = stateprep::make_uccsd_operator_pool(20, 10);

  ASSERT_FALSE(operators.empty());
  EXPECT_EQ(operators.size(), 875);
  expect_max_degree_less_than(operators, 20);
}

TEST(UCCSDOperatorPool, GeneratesCorrectOperators) {
  auto operators = stateprep::make_uccsd_operator_pool(4, 2);

  std::set<std::size_t> all_qubits{0, 1, 2, 3};
  for (auto &op : operators)
    op.canonicalize(all_qubits);

  std::vector<cudaq::spin_op> expected;
  expected.emplace_back(-0.500 * cudaq::spin_op::from_word("XZYI") +
                        +0.500 * cudaq::spin_op::from_word("YZXI"));
  expected.emplace_back(-0.500 * cudaq::spin_op::from_word("IXZY") +
                        +0.500 * cudaq::spin_op::from_word("IYZX"));
  expected.emplace_back(-0.125 * cudaq::spin_op::from_word("YYYX") +
                        -0.125 * cudaq::spin_op::from_word("YXXX") +
                        +0.125 * cudaq::spin_op::from_word("XXYX") +
                        -0.125 * cudaq::spin_op::from_word("YYXY") +
                        +0.125 * cudaq::spin_op::from_word("XYYY") +
                        +0.125 * cudaq::spin_op::from_word("XXXY") +
                        +0.125 * cudaq::spin_op::from_word("YXYY") +
                        -0.125 * cudaq::spin_op::from_word("XYXX"));

  ASSERT_EQ(expected.size(), operators.size());
  for (std::size_t i = 0; i < expected.size(); ++i)
    EXPECT_EQ(expected[i], operators[i]) << "Mismatch at index " << i;
}
